DOCSP-29490: Created Connection Troubleshooting Guide (#102)
mongodb/docs-csharp: source/connection-troubleshooting.txt

port is ``27017``:

.. code-block:: none
   :copyable: false

   Error: couldn't connect to server 127.0.0.1:27017

The following sections describe actions you can take to potentially resolve the
issue.

.. _csharp-troubleshooting-connection-string-port:

Check Your Connection String
----------------------------

Verify that the hostname and port number in the connection string are both
accurate. The default port value for a MongoDB instance is
``27017``, but you can configure MongoDB to communicate on another port.

.. _csharp-troubleshooting-connection-firewall:

Configure Your Firewall
-----------------------

Verify that the ports your MongoDB deployment listens on are not blocked by a
firewall on the same network. MongoDB uses port ``27017`` by default. To learn
more about the default ports MongoDB uses and how to change them, see
:manual:`Default MongoDB Port </reference/default-mongodb-port/>`.

.. warning::

   Do not open a port in your firewall unless you are sure it's the port
   used by your MongoDB deployment.

Authentication Error
~~~~~~~~~~~~~~~~~~~~

The {+driver-short+} can fail to connect to a MongoDB instance if

> the authorization is not configured correctly. If you are using ``SCRAM-SHA-256``

Reviewer: Q: in general, do we use authorization and authentication somewhat interchangeably?

Follow-up S depending on your answer:
```suggestion
the authentication mechanism is not configured correctly. If you are using ``SCRAM-SHA-256``
```

or ``SCRAM-SHA-1`` for authentication and the driver fails to connect, the
driver might raise an error message similar to one of the following messages:

.. code-block:: none
   :copyable: false

   Command failed with error 18 (AuthenticationFailed): 'Authentication
   failed.' on server <hostname>:<port>.

.. code-block:: none
   :copyable: false

   Authentication failed","attr":{"mechanism":"SCRAM-SHA-256","principalName":

>    "<user>","<auth databse>":"<user>","client":"127.0.0.1:2012",

Reviewer: S: make the placeholders follow the same formatting - either phrases with spaces, camelcase, etc

Author: I am not sure what you are indicating here. The place holders are  "<user>","<auth database>":"<user>" They all follow the same format (lowercase with spaces if two words).

The other terms are straight from the error message and are not placeholders.

Reviewer: ah okay thanks for clarifying!

   "result":"UserNotFound: Could not find user}}

.. code-block:: none
   :copyable: false

   connection() error occurred during connection handshake: auth error:
   sasl conversation error: unable to authenticate using mechanism
   "SCRAM-SHA-256": (AuthenticationFailed) Authentication failed.

The following sections describe actions you can take to potentially resolve the
issue.

.. _csharp-troubleshooting-connection-string-auth:

Check Your Connection String
----------------------------

An invalid connection string is the most common cause of authentication
issues when attempting to connect to MongoDB using connection strings and
``SCRAM-SHA-256`` or ``SCRAM-SHA-1``.

.. tip::

   For more information about connection strings,
   see :ref:`Connection URI <csharp-connection-uri>` in the Connection Guide.

If your connection string contains a username and password, ensure that they
are in the correct format. If the username or password includes any of the
following characters, they must be
`percent encoded <https://tools.ietf.org/html/rfc3986#section-2.1>`__:

.. code-block:: none

    : / ? # [ ] @

> The following example shows how to percent encode "#MyP@assword?":

Reviewer: I: On my machine, #MyP@assword creates a link to draft an email for some reason. Since we don't want users to experience unexpected linking out of our docs, could you change this to something that doesn't get picked up as a link/action?

.. code-block:: csharp

   Console.WriteLine(System.Web.HttpUtility.UrlEncode("#MyPaassword?"));

This results in the following output:

.. code-block:: none

>    %23MyPassword%3F

Author: I did. That is the new encoding. (no escape for the @ symbol and replaced with just an a)

>    %23MyPassword%3F

Reviewer: Then the first code block has an extra a - Console.WriteLine(System.Web.HttpUtility.UrlEncode("#**MyPaassword**?"));

.. _csharp-troubleshooting-connection-mongoclientsettings:

Verify the MongoClientSettings Properties
-----------------------------------------

You can use a ``MongoClientSettings`` object to configure the settings when
attempting to connect to a MongoDB deployment. You can use the ``Credential``
property to set authentication information. If this information is not correct,

> it will result in authentication errors when you attempt to connect to your

Reviewer: S: clarify "it" at the beginning of this line

MongoDB deployment.

.. _csharp-troubleshooting-connection-admin:

Verify the User Is in the Authentication Database
-------------------------------------------------

To successfully authenticate a connection by using a username and password with
``SCRAM-SHA-256`` or ``SCRAM-SHA-1``, the username must be defined in the
authentication database. The default authentication database is the ``admin``
database. To use a different database for authentication, specify the

> ``authSource`` in the connection string. The following example instructs the

Reviewer: ```suggestion
``authSource`` option in the connection string. The following example instructs the
```

Too Many Open Connections
~~~~~~~~~~~~~~~~~~~~~~~~~

The driver creates the following error message when it attempts to open a
connection, but it's reached the maximum number of connections:

.. code-block:: none
   :copyable: false

   connection refused because too many open connections

The following section describes a method that may help resolve the issue.

Check the Number of Connections
-------------------------------

If you need to create more open connections, increase ``MaxConnectionPoolSize``. For more
information about checking the number of connections, see
:ref:`Check the Number of Connections <csharp-troubleshooting-connection-number-connections>`
in the Error Sending Message section.

Timeout Error
~~~~~~~~~~~~~

When the network is not able to deliver a request from the driver to the server
quickly enough, it can time out. When this happens, you might receive an error message
similar to the following message:

.. code-block:: none
   :copyable: false

   timed out while checking out a connection from connection pool: context canceled

If you receive this error, try the following action to resolve the
issue.

Set connectTimeoutMS
--------------------

> The driver may hang when it's unable to establish a connection because it

Reviewer: S: the "it" at the end of this line is a little bit unclear. Clarify if it refers to the driver or the connection

takes too long attempting to reach unreachable replica set nodes. You can limit the
time the driver spends attempting to establish the connection by using the
``connectTimeMS`` setting. To learn more about this setting, see the
:manual:`Timeout Options </reference/connection-string/#timeout-options>` in

> the server manual.

Reviewer: ```suggestion
the Server manual.
```
